apps/audit/auditors.py

```diff
@@ -0,0 +1,45 @@
+import logging
+
+from django.core.exceptions import ValidationError
+from field_audit.auditors import BaseAuditor
+from field_audit.models import USER_TYPE_REQUEST
+
+from apps.users.models import CustomUser
+
+log = logging.getLogger("audit")
+
+
+class RequestAuditor(BaseAuditor):
+    """Auditor class for getting users from authenticated requests."""
+
+    def change_context(self, request):
+        if request is None:
+            return None
+        if request.user.is_authenticated:
+            username = get_request_username(request)
+            context = {
+                "user_type": USER_TYPE_REQUEST,
+                "username": username,
+            }
+            if username != request.user.username:
+                context["as_username"] = request.user.username
+            return context
+        # short-circuit the audit chain for not-None requests
+        return {}
+
+
+def get_request_username(request):
+    hijack_history = request.session.get("hijack_history", [])
+    if hijack_history:
+        if username := _get_hijack_username(hijack_history):
+            return username
+
+    return request.user.username
+
+
+def _get_hijack_username(hijack_history):
+    try:
+        pk = CustomUser._meta.pk.to_python(hijack_history[-1])
```

Comment:
Why this line?

Reply:
it's a 'safe' way to convert from the serialised value (string) to the python datatype (int)

Reply:
Maybe I'm missing an obvious scenario here, but when would it be unsafe? If users tamper with the session?

Reply:
It shouldn't be unsafe but we have to convert str to int somehow and this uses the model fields conversion instead of just going `int(...)`. Either way we still have to handle the error case/

Reply:
So is it to achieve the following?

![image](https://github.com/dimagi/open-chat-studio/assets/64970009/34804f20-b85b-4075-a3b2-23135751f3f9)

Reply:
No not really since that already works, it's just a validation step to make sure the `pk` is what we expect. We could probably remove it and Django would throw another error if the value wasn't an integer.